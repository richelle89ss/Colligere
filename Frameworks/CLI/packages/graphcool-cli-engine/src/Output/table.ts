import { Output } from './index'
import * as stripAnsi from 'strip-ansi'

export interface TableColumn<T> {
  key: string,
  label?: string,
  format: (value: string, row: string) => string,
  get: (row: T) => string,
  width: number
}

export interface TableOptions<T> {
  columns?: Array<TableColumn<T>>,
  after: (row: T, options: TableOptions<T>) => void,
  printRow: (row: any[]) => void,
  printHeader?: (row: any[]) => void
}

/**
 * Generates a Unicode table and feeds it into configured printer.
 *
 * Top-level arguments:
 *
 * @arg {Object[]} data - the records to format as a table.
 * @arg {Object} options - configuration for the table.
 *
 * @arg {Object[]} [options.columns] - Options for formatting and finding values for table columns.
 * @arg {function(string)} [options.headerAnsi] - Zero-width formattter for entire header.
 * @arg {string} [options.colSep] - Separator between columns.
 * @arg {function(row, options)} [options.after] - Function called after each row is printed.
 * @arg {function(string)} [options.printLine] - Function responsible for printing to terminal.
 * @arg {function(cells)} [options.printHeader] - Function to print header cells as a row.
 * @arg {function(cells)} [options.printRow] - Function to print cells as a row.
 *
 * @arg {function(row)|string} [options.columns[].key] - Path to the value in the row or function to retrieve the pre-formatted value for the cell.
 * @arg {function(string)} [options.columns[].label] - Header name for column.
 * @arg {function(string, row)} [options.columns[].format] - Formatter function for column value.
 * @arg {function(row)} [options.columns[].get] - Function to return a value to be presented in cell without formatting.
 *
 */
function table<T = { height?: number }>(out: Output, data: any[], options: TableOptions<T>) {
  const ary = require('lodash.ary')
  const defaults = require('lodash.defaults')
  const get = require('lodash.get')
  const identity = require('lodash.identity')
  const partial = require('lodash.partial')
  const property = require('lodash.property')
  const result = require('lodash.result')

  const defaultOptions = {
    colSep: '  ',
    after: () => {
      // noop
    },
    headerAnsi: identity,
    printLine: s => out.log(s),
    printRow(cells) {
      this.printLine(cells.join(this.colSep).trimRight())
    },
    printHeader(cells) {
      this.printRow(cells.map(ary(this.headerAnsi, 1)))
      this.printRow(cells.map(hdr => hdr.replace(/./g, 'â”€')))
    }
  }

  const colDefaults = {
    format: value => value ? value.toString() : '',
    width: 0,
    label() {
      return this.key.toString()
    },

    get (row) {
      const path = result(this, 'key')
      const value = !path ? row : get(row, path)
      return this.format(value, row)
    }
  }

  function calcWidth(cell) {
    const lines = stripAnsi(cell).split(/[\r\n]+/)
    const lineLengths = lines.map(property('length'))
    return Math.max.apply(Math, lineLengths)
  }

  function pad(str: string, length: number) {
    const visibleLength = stripAnsi(str).length
    const diff = length - visibleLength

    return str + ' '.repeat(Math.max(0, diff))
  }

  function render() {
    let columns: Array<TableColumn<T>> = options.columns || Object.keys((data[0] as any) || {}) as any

    if (typeof columns[0] === 'string') {
      columns = (columns as any).map(key => ({key}))
    }

    let defaultsApplied = false
    for (const row of data) {
      row.height = 1
      for (const col of columns) {
        if (!defaultsApplied) {
          defaults(col, colDefaults)
        }

        const cell = col.get(row)

        col.width = Math.max(
          result(col, 'label').length,
          col.width || 0,
          calcWidth(cell)
        )

        row.height = Math.max(
          row.height || 0,
          cell.split(/[\r\n]+/).length
        )
      }
      defaultsApplied = true
    }

    if (options.printHeader) {
      options.printHeader(columns.map((col) => {
        const label = result(col, 'label')
        return pad(label, col.width || label.length)
      }))
    }

    function getNthLineOfCell(n, row, col) {
      // TODO memoize this
      const lines = col.get(row).split(/[\r\n]+/)
      return pad(lines[n] || '', col.width)
    }

    for (const row of data) {
      for (let i = 0; i < (row.height || 0); i++) {
        const cells = columns.map(partial(getNthLineOfCell, i, row))
        options.printRow(cells)
      }
      options.after(row, options)
    }
  }

  defaults(options, defaultOptions)
  render()
}

module.exports = table
