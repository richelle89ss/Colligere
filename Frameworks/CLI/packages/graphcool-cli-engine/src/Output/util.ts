import chalk from 'chalk'

export function makePartsEnclodesByCharacterBold(str: string, character: string): string {
  const components = str.split(character)
  for(let i = 0; i < components.length; i++) {
    if (i % 2 === 1) {
      components[i] = chalk.bold(components[i])
    }
  }
  return components.join(chalk.bold(`\``))
}

export function regionEnumToOption(regionEnum: string): string {
  return regionEnum.toLowerCase().replace(/_/g, '-')
}
