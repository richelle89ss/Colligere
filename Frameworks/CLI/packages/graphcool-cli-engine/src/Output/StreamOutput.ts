import * as util from 'util'
import chalk from 'chalk'
import * as path from 'path'
import * as fs from 'fs-extra'
import { Output } from './'
import * as stripAnsi from 'strip-ansi'

export function logToFile (msg: string, logfile: string) {
  try {
    fs.mkdirpSync(path.dirname(logfile))
    fs.appendFileSync(logfile, stripAnsi(msg))
  } catch (err) { console.error(err) }
}

export default class StreamOutput {
  static startOfLine = true

  output = ''
  stream: NodeJS.WriteStream
  out: Output
  logfile?: string

  constructor (stream: NodeJS.WriteStream, output: Output) {
    this.out = output
    this.stream = stream
  }

  write (msg: string, options: {log?: boolean} = {}) {
    let {startOfLine} = (this.constructor as any)
    const log = options.log !== false
    if (log) {
      this.writeLogFile(msg, startOfLine)
    }
    // conditionally show timestamp if configured to display
    if (startOfLine && this.displayTimestamps) {
      msg = this.timestamp(msg)
    }
    if (this.out.mock) {
      this.output += msg
    } else {
      this.stream.write(msg)
    }
    startOfLine = msg.endsWith('\n')
  }

  timestamp (msg: string): string {
    return `[${new Date().toISOString()}] ${msg}`
  }

  log (data: string, ...args: any[]) {
    let msg = data ? util.format(data, ...args) : ''
    msg += '\n'
    this.out.action.pause(() => this.write(msg))
  }

  writeLogFile (msg: string, withTimestamp: boolean) {
    if (!this.logfile) {
      return
    }
    msg = withTimestamp ? this.timestamp(msg) : msg
    logToFile(msg, this.logfile)
  }

  get displayTimestamps (): boolean {
    const bin = this.out.config.bin.replace('-', '_').toUpperCase()
    const key = `${bin}_TIMESTAMPS`
    return ['1', 'true'].includes(process.env[key] || '')
  }
}
