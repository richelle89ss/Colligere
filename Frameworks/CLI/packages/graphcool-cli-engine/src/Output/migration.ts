import {
  MigrationActionType,
  MigrationErrorMessage,
  MigrationMessage,
} from '../types/common'
import figures = require('figures')
import chalk from 'chalk'
import { Output } from './index'
import { makePartsEnclodesByCharacterBold } from './util'
import * as groupBy from 'lodash.groupby'

export class MigrationPrinter {
  out: Output
  constructor(out: Output) {
    this.out = out
  }
  printMessages(migrationMessages: MigrationMessage[]) {
    // group types,
    const groupedByType = groupBy(migrationMessages, m => m.type)
    Object.keys(groupedByType).forEach(type => {
      const typeMessages = groupedByType[type]
      this.out.log('\n' + printType(type) + '\n')
      const groupedByName = groupBy(typeMessages, m => m.name.split('.')[0])
      Object.keys(groupedByName).forEach(name => {
        this.out.log(`  ${chalk.bold(name)}`)
        const nameMessages = groupedByName[name]
        nameMessages.forEach(this.printMigrationMessage, this)
      })
    })
    this.out.log('')
  }
  printErrors(errors: MigrationErrorMessage[]) {
    const groupedByType = groupBy(errors, e => e.type)
    Object.keys(groupedByType).forEach(type => {
      const typeErrors = groupedByType[type]
      this.out.log('\n  ' + chalk.bold(type))
      typeErrors.forEach(error => {
        const outputMessage = makePartsEnclodesByCharacterBold(
          error.description,
          `\``,
        )
        this.out.log(`    ${chalk.red(figures.cross)} ${outputMessage}`)
      })
    })
  }

  private printMigrationMessage(
    migrationMessage: MigrationMessage,
    index: number,
  ) {
    const actionType = this.getType(migrationMessage.description)
    const symbol = this.getSymbol(actionType)
    const description = makePartsEnclodesByCharacterBold(
      migrationMessage.description,
      '`',
    )
    const outputMessage = `   ${symbol} ${description}`
    this.out.log(outputMessage)

    migrationMessage.subDescriptions!.forEach((subMessage, index2) => {
      const actionType2 = this.getType(subMessage.description)
      const symbol2 = this.getSymbol(actionType2)
      const outputMessage2 = makePartsEnclodesByCharacterBold(
        subMessage.description,
        '`',
      )
      const lastLine = index2 === migrationMessage.subDescriptions!.length - 1
      const endSymbol = lastLine ? '└' : '├'
      this.out.log(`   ${endSymbol}─ ${symbol2}  ${outputMessage2}`)
    })
  }

  private getType(message: string): MigrationActionType {
    if (message.indexOf('create') >= 0) {
      return 'create'
    } else if (message.indexOf('update') >= 0) {
      return 'update'
    } else if (
      message.indexOf('delete') >= 0 ||
      message.indexOf('remove') >= 0
    ) {
      return 'delete'
    }
    return 'unknown'
  }

  private getSymbol(type: MigrationActionType): string {
    switch (type) {
      case 'create':
        return chalk.green('+')
      case 'delete':
        return chalk.red('-')
      case 'update':
        return chalk.blue('*')
      case 'unknown':
        return chalk.cyan('?')
    }
  }
}

const spaces = (n: number) => Array(n + 1).join(' ')

const printType = (type: string) =>
  chalk.bold.underline(type.replace(/\b\w/g, l => l.toUpperCase()) + 's')
