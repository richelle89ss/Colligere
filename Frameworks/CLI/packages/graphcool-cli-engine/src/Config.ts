import { RunOptions } from './types/common'
import * as path from 'path'
import * as os from 'os'
import * as fs from 'fs-extra'
import * as cuid from 'scuid'
import * as findUp from 'find-up'
import { Output } from './Output/index'
import * as yaml from 'js-yaml'
const debug = require('debug')('config')


export class Config {
  /**
   * Local settings
   */
  out: Output
  debug: boolean = Boolean(
    process.env.DEBUG && process.env.DEBUG!.includes('*'),
  )
  windows: boolean = false
  bin: string = 'graphcool'
  mock: boolean = true
  argv: string[] = process.argv.slice(1)
  commandsDir: string = path.join(__dirname, '../dist/commands')
  defaultCommand: string = 'help'
  userPlugins: boolean = false
  version: string = '1.3.11'
  name: string = 'graphcool'
  pjson: any = {
    name: 'cli-engine',
    version: '0.0.0',
    dependencies: {},
    'cli-engine': {
      defaultCommand: 'help',
    },
  }
  sharedClusters: string[] = ['shared-eu-west-1', 'shared-ap-northeast-1', 'shared-us-west-2']

  /**
   * Paths
   */
  cwd: string
  home: string
  root = path.join(__dirname, '..')

  definitionDir: string
  definitionPath: string | null
  localRCPath: string
  globalRCPath: string
  warnings: string[] = []

  /**
   * Urls
   */
  authUIEndpoint = process.env.ENV === 'DEV'
    ? 'https://dev.console.graph.cool/cli/auth'
    : 'https://console.graph.cool/cli/auth'
  backendAddr = process.env.ENV === 'DEV'
    ? 'https://dev.api.graph.cool'
    : 'https://api.graph.cool'
  systemAPIEndpoint = process.env.ENV === 'DEV'
    ? 'https://dev.api.graph.cool/system'
    : 'https://api.graph.cool/system'
  authEndpoint = process.env.ENV === 'DEV'
    ? 'https://cli-auth-api.graph.cool/dev'
    : 'https://cli-auth-api.graph.cool/prod'
  docsEndpoint = process.env.ENV === 'DEV'
    ? 'https://dev.graph.cool/docs'
    : 'https://www.graph.cool/docs'
  statusEndpoint = 'https://crm.graph.cool/prod/status'

  /**
   * consumer endpoints
   */
  simpleAPIEndpoint = process.env.ENV === 'DEV'
    ? 'https://dev.api.graph.cool/simple/v1/'
    : 'https://api.graph.cool/simple/v1/'
  relayAPIEndpoint = process.env.ENV === 'DEV'
    ? 'https://dev.api.graph.cool/relay/v1/'
    : 'https://api.graph.cool/relay/v1/'
  fileAPIEndpoint = process.env.ENV === 'DEV'
    ? 'https://dev.api.graph.cool/file/v1/'
    : 'https://api.graph.cool/file/v1/'
  subscriptionsEndpoint = process.env.ENV === 'DEV'
    ? 'wss://dev.subscriptions.graph.cool'
    : 'wss://subscriptions.graph.cool'


  /* tslint:disable-next-line */
  __cache = {}

  constructor(options?: RunOptions) {
    this.cwd = (options && options.cwd) || this.getCwd()
    this.home = (options && options.home) || this.getHome()
    debug(`CWD`, this.cwd)
    debug(`HOME`, this.home)
    this.setDefinitionPaths()
    this.setRCPaths()
    if (options) {
      this.readPackageJson(options)
    }
  }
  setLocal(host: string = 'http://localhost:60000') {
    this.backendAddr = host
    this.systemAPIEndpoint = host + '/system'
  }
  setOutput(out: Output) {
    this.out = out
    this.warnings.forEach(warning  => out.warn(warning))
    this.warnings = []
  }
  get arch(): string {
    return os.arch() === 'ia32' ? 'x86' : os.arch()
  }
  get platform(): string {
    return os.platform() === 'win32' ? 'windows' : os.platform()
  }
  get userAgent(): string {
    return `${this.name}/${this.version} (${this.platform}-${this
      .arch}) node-${process.version}`
  }
  get dirname() {
    return this.pjson['cli-engine'].dirname || this.bin
  }
  get cacheDir() {
    const x = dir(
      this,
      'cache',
      this.platform === 'darwin'
        ? path.join(this.home, 'Library', 'Caches')
        : null,
    )
    return x
  }
  get requireCachePath() {
    return path.join(this.cacheDir, '/.require-cache.json')
  }
  private readPackageJson(options: RunOptions) {
    this.mock = options.mock
    this.argv = options.argv || this.argv
    if (options.root) {
      this.root = options.root
      const pjsonPath = path.join(options.root, 'package.json')
      const pjson = fs.readJSONSync(pjsonPath)
      if (pjson && pjson['cli-engine']) {
        this.pjson = pjson
        this.version = pjson.version
      }
    }
  }
  private setRCPaths() {
    this.localRCPath = path.join(this.definitionDir, '.graphcoolrc')
    const homePath = path.join(this.home, '.graphcoolrc')
    debug(`homepath`, homePath)
    this.globalRCPath = homePath
    debug(`localRCPath`, this.localRCPath)
    debug(`globalRCPath`, this.globalRCPath)
  }
  private warn(msg: string) {
    this.warnings.push(msg)
  }
  private setDefinitionPaths() {
    const definitionPath = path.join(this.cwd, 'graphcool.yml')
    if (fs.pathExistsSync(definitionPath)) {
      this.definitionDir = this.cwd
      this.definitionPath = definitionPath
    } else {
      const found = findUp.sync('graphcool.yml', {cwd: this.cwd})
      this.definitionDir = found ? path.dirname(found) : this.cwd
      this.definitionPath = found || null
    }
    debug(`definitionDir`, this.definitionDir)
    debug(`definitionPath`, this.definitionPath)
  }
  private getCwd() {
    // get cwd
    let cwd = process.cwd()
    if (process.env.NODE_ENV === 'test') {
      cwd = path.join(os.tmpdir(), `${cuid()}/`)
      fs.mkdirpSync(cwd)
      debug('cwd', cwd)
    }
    return cwd
  }
  private getHome() {
    // get home
    let home = os.homedir() || os.tmpdir()
    if (process.env.NODE_ENV === 'test') {
      home = path.join(os.tmpdir(), `${cuid()}/`)
      fs.mkdirpSync(home)
      debug('home', home)
    }
    return home
  }
}

function dir(config: Config, category: string, d: string | null): string {
  const cacheKey = `dir:${category}`
  const cache = config.__cache[cacheKey]
  if (cache) {
    return cache
  }
  d =
    d ||
    path.join(
      config.home,
      category === 'data' ? '.local/share' : '.' + category,
    )
  if (config.windows) {
    d = process.env.LOCALAPPDATA || d
  }
  d = process.env.XDG_DATA_HOME || d
  d = path.join(d, config.dirname)
  fs.mkdirpSync(d)
  config.__cache[cacheKey] = d
  return d
}
