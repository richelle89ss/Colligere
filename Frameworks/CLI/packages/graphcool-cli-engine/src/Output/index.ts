import { Config } from '../Config'
import StreamOutput, { logToFile } from './StreamOutput'
import chalk, { Chalk } from 'chalk'
import * as styles from 'ansi-styles'
import { ActionBase, shouldDisplaySpinner } from './actions/ActionBase'
import Prompter, { PromptOptions } from './Prompter'
import * as supports from 'supports-color'
import { SpinnerAction } from './actions/SpinnerAction'
import { SimpleAction } from './actions/SimpleAction'
import * as path from 'path'
import * as util from 'util'
import { errtermwidth } from './actions/screen'
import { TableOptions } from './table'
import ExitError from '../errors/ExitError'
import * as inquirer from 'graphcool-inquirer'
import { MigrationPrinter } from './migration'
import * as treeify from 'treeify'
import * as dirTree from 'directory-tree'
import * as marked from 'marked'
import * as TerminalRenderer from 'marked-terminal'
import * as Charm from 'charm'
import { padEnd, repeat, set, uniqBy, values } from 'lodash'
import { Project } from '../types/common'
import { Targets } from '../types/rc'

marked.setOptions({
  renderer: new TerminalRenderer(),
})

export const CustomColors = {
  // map gray -> dim because it's not solarized compatible
  supports,
  gray: (s: string) => chalk.dim(s),
  grey: (s: string) => chalk.dim(s),
  attachment: (s: string) => chalk.cyan(s),
  addon: (s: string) => chalk.yellow(s),
  configVar: (s: string) => chalk.green(s),
  release: (s: string) => chalk.blue.bold(s),
  cmd: (s: string) => chalk.green.bold(s),
  app: (s: string) => CustomColors.graphcool(`⬢ ${s}`),
  graphcool: (s: string) => {
    if (!CustomColors.supports) {
      return s
    }
    const has256 =
      CustomColors.supports.has256 ||
      (process.env.TERM || '').indexOf('256') !== -1
    return has256
      ? '\u001b[38;5;104m' + s + styles.reset.open
      : chalk.magenta(s)
  },
}

function wrap(msg: string): string {
  const linewrap = require('@heroku/linewrap')
  return linewrap(6, errtermwidth, {
    skipScheme: 'ansi-color',
    skip: /^\$ .*$/,
  })(msg)
}

function bangify(msg: string, c: string): string {
  const lines = msg.split('\n')
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i]
    lines[i] = ' ' + c + line.substr(2, line.length)
  }
  return lines.join('\n')
}

function extractMessage (response): string {
  try {
    return response.errors![0].message
  } catch (e) {
    return `GraphQL Error (Code: ${response.status})`
  }
}

const arrow = process.platform === 'win32' ? '!' : '▸'

export class Output {
  mock: boolean = true
  config: Config
  action: ActionBase
  stdout: StreamOutput
  stderr: StreamOutput
  prompter: Prompter
  prompt: any
  migration: MigrationPrinter
  charm: any

  constructor(config: Config) {
    this.config = config
    this.mock = config.mock
    this.stdout = new StreamOutput(process.stdout, this)
    this.stderr = new StreamOutput(process.stderr, this)
    this.action = shouldDisplaySpinner(this)
      ? new SpinnerAction(this)
      : new SimpleAction(this)
    if (this.mock) {
      chalk.enabled = false
      CustomColors.supports = false
    }
    this.prompter = new Prompter(this)
    this.prompt = inquirer.createPromptModule({ output: process.stdout })
    this.migration = new MigrationPrinter(this)
    this.charm = Charm()
    this.charm.pipe(process.stdout)
  }

  get color(): any {
    return new Proxy(chalk, {
      get: (chalkProxy, name) => {
        if (CustomColors[name]) {
          return CustomColors[name]
        }
        return chalkProxy[name]
      },
    })
  }

  log(data, ...args: any[]) {
    this.stdout.log(data, ...args)
  }

  getStyledJSON(obj: any, subtle: boolean = false) {
    const json = JSON.stringify(obj, null, 2)
    if (chalk.enabled) {
      const cardinal = require('cardinal')
      const theme = require(subtle ? './subtle' : 'cardinal/themes/jq')
      return cardinal.highlight(json, { json: true, theme })
    } else {
      return json
    }
  }

  async done(...rest: void[]): Promise<void> {
    this.action.stop()
  }

  debug(obj: string) {
    if (this.config.debug) {
      this.action.pause(() => console.log(obj))
    }
  }

  get errlog(): string {
    return path.join(this.config.cacheDir, 'error.log')
  }

  get autoupdatelog(): string {
    return path.join(this.config.cacheDir, 'autoupdate.log')
  }

  error(err: Error | string, exitCode: number | false = 1) {
    if (
      (this.mock && typeof err !== 'string' && exitCode !== false) ||
      process.env.NODE_ENV === 'test'
    ) {
      throw err
    }
    try {
      if (typeof err === 'string') {
        err = new Error(err)
      }
      this.logError(err)
      if (this.action.task) {
        this.action.stop(this.color.bold.red('!'))
      }
      if (this.config.debug) {
        this.stderr.log(err.stack || util.inspect(err))
      } else {
        this.stderr.log(
          this.isGraphQLError(err) ? this.getGraphQLErrorMessage(err) :
          bangify(wrap(this.getErrorMessage(err)), this.color.red(arrow)),
        )
        const instruction = (process.env.SHELL && process.env.SHELL!.endsWith('fish')) ? '$ set -x DEBUG "*"' : '$ export DEBUG="*"'
        this.stderr.log(
          `\nGet in touch if you need help: https://www.graph.cool/forum
To get more detailed output, run ${chalk.dim(instruction)}`,
        )
      }
    } catch (e) {
      console.error('error displaying error')
      console.error(e)
      console.error(err)
    }
    if (exitCode !== false) {
      this.exit(exitCode)
    }
  }

  isGraphQLError(err) {
    return err.message && err.request && err.response
  }

  warn(err: Error | string, prefix?: string) {
    this.action.pause(() => {
      try {
        prefix = prefix ? `${prefix} ` : ''
        err = typeof err === 'string' ? new Error(err) : err
        this.logError(err)
        if (this.config.debug) {
          if (process.stderr.write(`WARNING: ${prefix}`)) {
            this.stderr.log(err.stack || util.inspect(err))
          }
        } else {
          this.stderr.log(
            bangify(
              wrap(prefix + this.getErrorMessage(err)),
              this.color.yellow(arrow),
            ) + '\n',
          )
        }
      } catch (e) {
        console.error('error displaying warning')
        console.error(e)
        console.error(err)
      }
    }, this.color.bold.yellow('!'))
  }

  getErrorPrefix(fileName: string, type: 'error' | 'warning' = 'error') {
    const method = type === 'error' ? 'red' : 'yellow'
    return chalk[method](`[${type.toUpperCase()}] in ${chalk.bold(fileName)}: `)
  }

  logError(err: Error | string) {
    logToFile(util.inspect(err) + '\n', this.errlog)
  }

  printMarkdown(markdown: string) {
    this.log(
      marked(markdown)
        .split('\n')
        .map(l => `   ${l}`)
        .join('\n'),
    )
  }

  oldprompt(name: string, options?: PromptOptions) {
    return this.prompter.prompt(name, options)
  }

  table<T = { height?: number }>(data: T[], options: TableOptions<T>) {
    const table = require('./table')
    return table(this, data, options)
  }

  exit(code: number = 0) {
    if (this.config.debug) {
      console.error(`Exiting with code: ${code}`)
    }
    if (this.mock) {
      throw new ExitError(code)
    } else {
      process.exit(code)
    }
  }

  filesTree(files: string[]) {
    const tree = filesToTree(files)
    const printedTree = treeify.asTree(tree, true)
    this.log(chalk.dim(printedTree.split('\n').join('\n')))
  }

  tree(dirPath: string, padding = false) {
    const tree = dirTree(dirPath)
    const convertedTree = treeConverter(tree)
    const printedTree = treeify.asTree(convertedTree, true)
    this.log(
      chalk.blue(
        printedTree
          .split('\n')
          .map(l => (padding ? '  ' : '') + l)
          .join('\n'),
      ),
    )
  }

  up(y: number = 1) {
    for (let i = 0; i < y; i++) {
      this.charm.delete('line', 1)
      this.charm.up(1)
    }
  }

  printPadded(
    arr1: string[][],
    spaceLeft: number = 0,
    spaceBetween: number = 1,
    header?: string[],
  ) {
    const inputRows = arr1
    if (header) {
      inputRows.unshift(header)
    }
    const leftCol = inputRows.map(a => a[0])
    const rightCol = inputRows.map(a => a[1])
    const maxLeftCol = leftCol.reduce(
      (acc, curr) => Math.max(acc, curr.length),
      -1,
    )
    const maxRightCol = rightCol.reduce(
      (acc, curr) => Math.max(acc, curr.length),
      -1,
    )
    const paddedLeftCol = leftCol.map(
      v => repeat(' ', spaceLeft) + padEnd(v, maxLeftCol + spaceBetween),
    )

    const rows = paddedLeftCol.map((l, i) => l + arr1[i][1])

    if (header) {
      const divider = `${repeat('─', maxLeftCol)}${repeat(
        ' ',
        spaceBetween,
      )}${repeat('─', maxRightCol)}`
      rows.splice(1, 0, divider)
    }

    return rows.join('\n')
  }

  printServices = (
    targets: Targets,
    projects: Project[],
    onlyLocal: boolean = true,
  ) => {
    const uniqTargetKeys = uniqBy(Object.keys(targets), key => targets[key].id)
    if (onlyLocal) {
      return this.printPadded(
        uniqTargetKeys.map(key => {
          const { id, cluster } = targets[key]
          const project = projects.find(p => p.id === id)
          const output = `${cluster}/${id}`
          const serviceName = project ? project.name : key
          return [serviceName + '  ', output + '  ']
        }),
        0,
        1,
        ['Service Name', 'Cluster / Service ID'],
      )
    } else {
      const filteredProjects = projects.filter(
        p => !values(targets).find(t => p.id === t.id),
      )
      return this.printPadded(
        uniqTargetKeys
          .map(key => {
            const { id, cluster } = targets[key]
            const project = projects.find(p => p.id === id)
            const output = `${cluster}/${id}`
            const serviceName = project ? project.name : key
            return [serviceName, output]
          })
          .concat(
            filteredProjects.map(p => {
              return [
                p.name,
                `shared-${p.region.toLowerCase().replace(/_/g, '-')}/${p.id}`,
              ]
            }),
          )
          .map(l => [l[0] + '  ', l[1] + '  ']),
        0,
        1,
        ['Service Name', 'Cluster / Service ID'],
      )
    }
  }
  getGraphQLErrorMessage(err: any) {
    if (this.mock) {
      return `\n${chalk.bold.red('ERROR: ' + extractMessage(err.response))}\n\n`
        + chalk.bold('Request')
        + this.getStyledJSON(err.request, true)
        + chalk.bold('Response')
        + this.getStyledJSON(err.response, true)
    } else {
      return `\n${chalk.bold.red('ERROR: ' + extractMessage(err.response))}\n\n` + this.getStyledJSON(err.response, true)
    }
  }
  getErrorMessage(err: any): string {
    let message

    if (err.body) {
      // API error
      if (err.body.message) {
        message = util.inspect(err.body.message)
      } else if (err.body.error) {
        message = util.inspect(err.body.error)
      }
    }
    // Unhandled error
    if (err.message && err.code) {
      message = `${chalk.bold(util.inspect(err.code))}: ${err.message}`
    } else if (err.message) {
      message = err.message
    }
    return message || util.inspect(err)
  }

}

function treeConverter(tree) {
  if (!tree.children) {
    return tree.name
  }
  return tree.children.reduce((acc, curr) => {
    if (!curr.children) {
      return { ...acc, [curr.name]: null }
    }
    return { ...acc, [curr.name]: treeConverter(curr) }
  }, {})
}

function filesToTree(files: string[]) {
  const fileNames = files.map(l => {
    if (l.startsWith('./')) {
      return l.slice(2)
    }

    return l
  })

  const obj = {}

  fileNames.forEach(fileName => {
    const setPath = fileName.split('/')
    set(obj, setPath, null)
  })

  return obj
}
