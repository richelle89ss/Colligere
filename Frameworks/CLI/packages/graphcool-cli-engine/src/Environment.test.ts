import {Environment} from './Environment'
import {Output} from './Output'
import {Config} from './Config'

// lets write some tests

function makeEnvironment() {
  const config = new Config()
  const out = new Output(config)
  return new Environment(out, config)
}

describe('environment', () => {
  test('should work without global RC', async () => {
    const env = makeEnvironment()
    const localFile = `
platformToken: 'secret-token'
targets:
  dev: local/asdasd123
clusters:
  local:
    host: http://localhost:60000
    token: asdf
`

    await env.loadRCs(localFile, null)
    expect(env.rc).toMatchSnapshot()
  })
  test('should resolve a target alias refering before', async () => {
    const env = makeEnvironment()
    const localFile = `
platformToken: 'secret-token'
targets:
  dev: local/asdasd123
  test-alias: dev
clusters:
  local:
    host: http://localhost:60000
    token: asdf
`

    await env.loadRCs(localFile, null)
    expect(env.rc).toMatchSnapshot()
  })
  test('should resolve a target alias refering after', async () => {
    const env = makeEnvironment()
    const localFile = `
platformToken: 'secret-token'
targets:
  test-alias: dev
  dev: local/asdasd123
clusters:
  local:
    host: http://localhost:60000
    token: asdf
`

    await env.loadRCs(localFile, null)
    expect(env.rc).toMatchSnapshot()
  })
//
//   test('throws when cluster does not exist', async () => {
//     const env = makeEnvironment()
//     const localFile = `
// platformToken: 'secret-token'
// targets:
//   dev: local/asdasd123
// `
//     env.loadRCs(localFile, null)
//     await expect(env.out.stderr.output).toMatch(/Could not find cluster local defined for target dev in/)
//   })

  test('should access targets from the global RC', async () => {
    const env = makeEnvironment()
    const localFile = `
platformToken: 'secret-token'
targets:
  test-alias: dev
  dev: local/asdasd123
`
    const globalFile = `
clusters:
  local:
    host: http://localhost:60000
    token: asdf
    `

    await env.loadRCs(localFile, globalFile)
    expect(env.rc).toMatchSnapshot()
  })
  test('should resolve target from the global RC', async () => {
    const env = makeEnvironment()
    const localFile = `
platformToken: 'secret-token'
targets:
  test-alias: dev
`
    const globalFile = `
targets:
  dev: local/asdasd123
clusters:
  local:
    host: http://localhost:60000
    token: asdf
    `

    await env.loadRCs(localFile, globalFile)
    expect(env.rc).toMatchSnapshot()
  })
  test('should override global target', async () => {
    const env = makeEnvironment()
    const localFile = `
platformToken: 'secret-token'
targets:
  dev: shared-eu-west-1/asdfasdf
`
    const globalFile = `
targets:
  dev: local/asdasd123
clusters:
  local:
    host: http://localhost:60000
    token: asdf
    `

    await env.loadRCs(localFile, globalFile)
    expect(env.rc).toMatchSnapshot()
  })
})
