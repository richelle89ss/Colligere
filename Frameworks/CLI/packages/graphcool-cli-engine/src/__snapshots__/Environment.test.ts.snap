// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`environment should access targets from the global RC 1`] = `
Object {
  "clusters": Object {
    "local": Object {
      "host": "http://localhost:60000",
      "token": "asdf",
    },
  },
  "platformToken": "secret-token",
  "targets": Object {
    "dev": Object {
      "cluster": "local",
      "id": "asdasd123",
    },
    "test-alias": Object {
      "cluster": "local",
      "id": "asdasd123",
    },
  },
}
`;

exports[`environment should override global target 1`] = `
Object {
  "clusters": Object {
    "local": Object {
      "host": "http://localhost:60000",
      "token": "asdf",
    },
  },
  "platformToken": "secret-token",
  "targets": Object {
    "dev": Object {
      "cluster": "shared-eu-west-1",
      "id": "asdfasdf",
    },
  },
}
`;

exports[`environment should resolve a target alias refering after 1`] = `
Object {
  "clusters": Object {
    "local": Object {
      "host": "http://localhost:60000",
      "token": "asdf",
    },
  },
  "platformToken": "secret-token",
  "targets": Object {
    "dev": Object {
      "cluster": "local",
      "id": "asdasd123",
    },
    "test-alias": Object {
      "cluster": "local",
      "id": "asdasd123",
    },
  },
}
`;

exports[`environment should resolve a target alias refering before 1`] = `
Object {
  "clusters": Object {
    "local": Object {
      "host": "http://localhost:60000",
      "token": "asdf",
    },
  },
  "platformToken": "secret-token",
  "targets": Object {
    "dev": Object {
      "cluster": "local",
      "id": "asdasd123",
    },
    "test-alias": Object {
      "cluster": "local",
      "id": "asdasd123",
    },
  },
}
`;

exports[`environment should resolve target from the global RC 1`] = `
Object {
  "clusters": Object {
    "local": Object {
      "host": "http://localhost:60000",
      "token": "asdf",
    },
  },
  "platformToken": "secret-token",
  "targets": Object {
    "dev": Object {
      "cluster": "local",
      "id": "asdasd123",
    },
    "test-alias": Object {
      "cluster": "local",
      "id": "asdasd123",
    },
  },
}
`;

exports[`environment should work without global RC 1`] = `
Object {
  "clusters": Object {
    "local": Object {
      "host": "http://localhost:60000",
      "token": "asdf",
    },
  },
  "platformToken": "secret-token",
  "targets": Object {
    "dev": Object {
      "cluster": "local",
      "id": "asdasd123",
    },
  },
}
`;
